scipy/sparse/_construct.py

```diff
@@ -1048,45 +1173,147 @@ def random(m, n, density=0.01, format='coo', dtype=None,
         tp = np.int64
 
     if mn > np.iinfo(tp).max:
-        msg = """\
-Trying to generate a random sparse matrix such as the product of dimensions is
-greater than %d - this is not supported on this machine
-"""
+        msg = (
+            "Trying to generate a random sparse matrix such that the product of "
+            "dimensions is\n greater than %d is not supported on this machine\n"
+        )
         raise ValueError(msg % np.iinfo(tp).max)
 
     # Number of non zero values
     k = int(round(density * m * n))
 
     random_state = check_random_state(random_state)
 
-    if data_rvs is None:
+    if data_rng is None:
         if np.issubdtype(dtype, np.integer):
-            def data_rvs(n):
+            def data_rng(n):
                 return rng_integers(random_state,
                                     np.iinfo(dtype).min,
                                     np.iinfo(dtype).max,
                                     n,
                                     dtype=dtype)
         elif np.issubdtype(dtype, np.complexfloating):
-            def data_rvs(n):
+            def data_rng(n):
                 return (random_state.uniform(size=n) +
                         random_state.uniform(size=n) * 1j)
         else:
-            data_rvs = partial(random_state.uniform, 0., 1.)
+            data_rng = partial(random_state.uniform, 0., 1.)
 
     ind = random_state.choice(mn, size=k, replace=False)
```

Comment:
This raises exceptions or gives duplicate entries (because `replace=False` only applies across 1 axis instead of all axes). I think a fast unravel-er like `np.unravel_index` could be worthwhile:
```python
    indices = np.unravel_index(random_state.choice(np.prod(shape), size=k, replace=False), shape=shape)
```

Reply:
That's kind of weird, right, one would expect it to be totally feasible to sample with replacement across N dimensions. Perhaps a feature we should add to NumPy!

Reply:
https://mail.python.org/archives/list/numpy-discussion@python.org/thread/O7CBSJU5I2PXMOBZC7MQ7FKFSO33E3FS/

With thanks to @rkern.

If we're happy with the uint64 upper limited, going with the `choice` solution works. At least we have options, should this become an issue later.

Reply:
The current code uses dtype `np.intc`. I don't know if people are running into that limit or not...
The choice code is something like this:
```python
     mn = np.prod(shape)
     size = int(round(density * mn))
     raveled_ind = rng.choice(np.prod(shape), size=size, replace=False).astype(dtype)
     indices = np.unravel_index(raveled_ind, shape=shape)
```
The integers method is slower maybe somewhat because of repacking after being generated.
```python
     mn = np.prod(shape)
     size = int(round(density * mn))
     seen = set()
     while len(seen) < size:
         dsize = size - len(seen)
         seen.update(map(tuple, rng.integers(shape, size=(dsize, ndim), dtype=dtype)))
     indices = np.array(list(seen), dtype=dtype).T
```
Nice versions for ints larger than 2**64 but that's probably not needed I'm guessing.
:)

Reply:
No, but probably worth ensuring that prod(shape) fits in intc. uint always makes more sense, int64 may make more sense sometimes?

Reply:
Right -- the max value has to fit...

The `choice` version is about 30 times faster in my tests for `size = 200-20000`
(when `np.prod(shape)` is small enough that it can work). 
So I propose a hybrid method depending on whether `math.prod(shape) < np.iinfo('uint64').max`.

Also, are we supposed to use `scipy.utils.rng_integers` instead of `rng.integers` so it works with a `random_state` input?

Reply:
FYI, `choice()` is restricted to `mn` being an `int64`.

Reply:
I've implemented a hybrid approach dependent on the size of `math.prod(shape)`. I used the `int64` maximum in the code rather than `uint64` because that matches scipy/sparse's choice for `idx_dtype` throughout (either int32 or int64). I'm not sure of the history there.